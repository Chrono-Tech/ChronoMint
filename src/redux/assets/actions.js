import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const ASSETS_LIST = 'assets/LIST'

export const listAssets = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getAssetsManagerDAO()
  dispatch({type: ASSETS_LIST, list: await dao.getList()})
}
