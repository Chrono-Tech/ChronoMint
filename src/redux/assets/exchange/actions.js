import type ExchangeManagerDAO from 'dao/ExchangeManagerDAO'
import type ExchangeModel from 'models/ExchangeModel'
import type TokenModel from 'models/TokenModel'
import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const EXCHANGE_MANAGER_LIST = 'exchangeManager/LIST'
export const EXCHANGE_MANAGER_SET = 'exchangeManager/SET'
export const EXCHANGE_MANAGER_REMOVE = 'exchangeManager/REMOVE'

const getDAO = async (): Promise<ExchangeManagerDAO> => await contractsManagerDAO.getExchangeManagerDAO()

export const listExchanges = () => async (dispatch) => {
  const dao = await getDAO()
  dispatch({type: EXCHANGE_MANAGER_LIST, list: await dao.getList()})
}

export const addExchange = (address) => async () => {
  const exchange = new ExchangeModel({address, isFetching: true})
  dispatch({type: EXCHANGE_MANAGER_SET, exchange})
  try {
    const dao = await getDAO()
    await dao.addExchange(address)
  } catch (e) {
    // no rollback
  }
}

export const createExchange = (token: TokenModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.createExchange(token)
  } catch (e) {
    // no rollback
  }
}

export const editExchange = (exchange: ExchangeModel, newExchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.editExchange(exchange, newExchange)
  } catch (e) {
    // no rollback
  }
}

export const removeExchange = (exchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.removeExchange(exchange)
  } catch (e) {
    // no rollback
  }
}

// TODO @bshevchenko: MINT-230 AssetsManager
// export const addOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.addOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
//
// export const removeOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.removeOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
