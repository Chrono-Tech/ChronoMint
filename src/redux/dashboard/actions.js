import LHTProxyDAO from '../../dao/LHTProxyDAO'
import UserStorageDAO from '../../dao/UserStorageDAO'

import {
  DASHBOARD_TOTAL_LHT_FETCH,
  DASHBOARD_TOTAL_LHT,
  DASHBOARD_TOTAL_MEMBERS_FETCH,
  DASHBOARD_TOTAL_MEMBERS
} from './reducer'

export const updateTotalLHT = () => (dispatch) => {
  dispatch({type: DASHBOARD_TOTAL_LHT_FETCH})
  return LHTProxyDAO.totalSupply()
    .then(balance => {
      dispatch({type: DASHBOARD_TOTAL_LHT, payload: balance})
    })
}

export const updateTotalMembers = () => (dispatch) => {
  dispatch({type: DASHBOARD_TOTAL_MEMBERS_FETCH})
  return UserStorageDAO.usersTotal()
    .then(number => {
      dispatch({type: DASHBOARD_TOTAL_MEMBERS, payload: number})
    })
}
