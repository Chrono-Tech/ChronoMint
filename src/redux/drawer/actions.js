/**
 * Copyright 2017â€“2018, LaborX PTY
 * Licensed under the AGPL Version 3 license.
 */

import {
  DRAWER_HIDE,
  DRAWER_TOGGLE,
} from './constants'

export const drawerToggle = () => (dispatch) => {
  dispatch({ type: DRAWER_TOGGLE })
}
export const drawerHide = () => (dispatch) => {
  dispatch({ type: DRAWER_HIDE })
}
