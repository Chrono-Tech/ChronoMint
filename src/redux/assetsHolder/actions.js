import { DUCK_SESSION } from '@/redux/session/actions'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import TIMEHolderDAO from 'dao/TIMEHolderDAO'
import Amount from 'models/Amount'
import ApprovalNoticeModel from 'models/notices/ApprovalNoticeModel'
import TokenModel from 'models/tokens/TokenModel'
import { DUCK_TOKENS } from 'redux/tokens/actions'
import tokenService, { EVENT_NEW_TOKEN } from 'services/TokenService'

export const DUCK_ASSETS_HOLDER = 'assetsHolder'

export const TIME_HOLDER_INIT = 'timeHolder/INIT'
export const TIME_HOLDER_TIME_ADDRESS = 'timeHolder/timeAddress'
export const TIME_HOLDER_WALLET_ADDRESS = 'timeHolder/timeHolderWalletAddress'
export const TIME_HOLDER_ADDRESS = 'timeHolder/timeHolderAddress'

let timeHolderDAO: TIMEHolderDAO = null

const subscribeOnTokens = (token: TokenModel) => async (dispatch, getState) => {
  if (token.symbol() !== 'TIME') {
    return
  }

  dispatch({ type: TIME_HOLDER_TIME_ADDRESS, address: token.address() })

  // init
  const [ timeHolderAddress, timeHolderWalletAddress ] = await Promise.all([
    timeHolderDAO.getAddress(),
    timeHolderDAO.getWalletAddress(),
  ])

  dispatch({ type: TIME_HOLDER_ADDRESS, address: timeHolderAddress })
  dispatch({ type: TIME_HOLDER_WALLET_ADDRESS, address: timeHolderWalletAddress })

  // TODO @dkchv: review again
  const contractNames = {}
  contractNames[ timeHolderAddress ] = 'TIME Holder'
  ApprovalNoticeModel.setContractNames(contractNames)

  // update allowance
  const timeDAO = tokenService.getDAO(token.id())
  const { account } = getState().get(DUCK_SESSION)
  const [ timeHolderAllowance, timeHolderWalletAllowance ] = await Promise.all([
    timeDAO.getAccountAllowance(timeHolderAddress, account),
    timeDAO.getAccountAllowance(timeHolderWalletAddress, account),
  ])

  console.log('--assetHolder: ', +timeHolderAllowance, +timeHolderWalletAllowance)

  // token = token
  //   .setAllowance(timeHolderAddress, timeHolderAllowance)
  //   .setAllowance(timeHolderWalletAddress, timeHolderWalletAllowance)
}

export const initAssetsHolder = () => async (dispatch, getState) => {
  if (getState().get(DUCK_ASSETS_HOLDER).isInited()) {
    return
  }
  dispatch({ type: TIME_HOLDER_INIT, inInited: true })

  const { account } = getState().get(DUCK_SESSION)
  timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const [ assets, deposit ] = await Promise.all([
    timeHolderDAO.getSharesContract(),
    timeHolderDAO.getDeposit(account),
  ])
  console.log('--deposit', assets, +deposit)

  // subscribe to tokens
  const callback = (token) => dispatch(subscribeOnTokens(token))
  tokenService.on(EVENT_NEW_TOKEN, callback)
  // fetch for existing tokens
  const tokens = getState().get(DUCK_TOKENS)
  tokens.list().forEach(callback)
}

export const depositAsset = (amount: Amount, token: TokenModel) => async (dispatch) => {
  // const amountBN = new BigNumber(amount)

  // dispatch(balanceMinus(amountBN, token))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.deposit(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('deposit error', e.message)
  } finally {
    // compensation for update in watchTransfer
    // dispatch(balancePlus(amount, token))
  }
}

export const withdrawAsset = (amount: Amount, token: TokenModel) => async () => {
  // const amountBN = new BigNumber(amount)

  // dispatch(depositMinus(amountBN))
  const assetDAO = tokenService.getDAO(token)

  try {
    await timeHolderDAO.withdraw(amount, assetDAO)
  } catch (e) {
    // eslint-disable-next-line
    console.error('withdraw error', e.message)
  } finally {
    // dispatch(depositPlus(amountBN))
  }
}
