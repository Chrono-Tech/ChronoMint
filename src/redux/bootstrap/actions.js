import ls from 'utils/LocalStorage'
import { LOCAL_ID } from 'network/settings'

import {
  checkMetaMask, checkLocalSession, restoreLocalSession, createNetworkSession,
  checkTestRPC
} from '../network/actions'
import { login } from '../session/actions'

export const bootstrap = (relogin = true) => async (dispatch) => {
  dispatch(checkMetaMask())
  dispatch(checkTestRPC())

  if (!relogin) {
    return
  }

  const localAccount = ls.getLocalAccount()
  const isPassed = await dispatch(checkLocalSession(localAccount))
  if (isPassed) {
    await dispatch(restoreLocalSession(localAccount))
    dispatch(createNetworkSession(localAccount, LOCAL_ID, LOCAL_ID))
    dispatch(login(localAccount))
  } else {
    // eslint-disable-next-line
    console.warn('Can\'t restore local session')
  }
}
