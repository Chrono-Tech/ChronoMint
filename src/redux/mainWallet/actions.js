import BigNumber from 'bignumber.js'
import { TXS_PER_PAGE } from 'dao/AbstractTokenDAO'
import assetDonatorDAO from 'dao/AssetDonatorDAO'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import ethereumDAO from 'dao/EthereumDAO'
import Immutable from 'immutable'
import { bccProvider, btcProvider } from 'Login/network/BitcoinProvider'
import { nemProvider } from 'Login/network/NemProvider'
import ApprovalNoticeModel from 'models/notices/ApprovalNoticeModel'
import TransferNoticeModel from 'models/notices/TransferNoticeModel'
import type ProfileModel from 'models/ProfileModel'
import TokenModel from 'models/TokenModel'
import type TxModel from 'models/TxModel'
import { notify } from 'redux/notifier/actions'
import { DUCK_SESSION } from 'redux/session/actions'
import tokenService, { EVENT_NEW_TOKEN } from 'services/TokenService'
import { addMarketToken } from '../market/action'

export const DUCK_MAIN_WALLET = 'mainWallet'

export const WALLET_TOKENS_FETCH = 'mainWallet/TOKENS_FETCH'
export const WALLET_TOKENS = 'mainWallet/TOKENS'
export const WALLET_BALANCE = 'mainWallet/BALANCE'
export const WALLET_BALANCE_SET = 'mainWallet/BALANCE_SET'
export const WALLET_ALLOWANCE = 'mainWallet/ALLOWANCE'
export const WALLET_TIME_DEPOSIT = 'mainWallet/TIME_DEPOSIT'
export const WALLET_TIME_ADDRESS = 'mainWallet/TIME_ADDRESS'
export const WALLET_BTC_ADDRESS = 'mainWallet/BTC_ADDRESS'
export const WALLET_BCC_ADDRESS = 'mainWallet/BCC_ADDRESS'
export const WALLET_NEM_ADDRESS = 'mainWallet/NEM_ADDRESS'
export const WALLET_TRANSACTIONS_FETCH = 'mainWallet/TRANSACTIONS_FETCH'
export const WALLET_TRANSACTION = 'mainWallet/TRANSACTION'
export const WALLET_TRANSACTIONS = 'mainWallet/TRANSACTIONS'
export const WALLET_IS_TIME_REQUIRED = 'mainWallet/IS_TIME_REQUIRED'

export const ETH = ethereumDAO.getSymbol()
export const TIME = 'TIME'
export const LHT = 'LHT'

export const setBalance = (token: TokenModel, amount: BigNumber) => ({
  type: WALLET_BALANCE_SET,
  token,
  amount,
})

export const updateBalance = (token: TokenModel, isCredited, amount: BigNumber) => ({
  type: WALLET_BALANCE,
  token,
  isCredited,
  amount,
})

export const balancePlus = (amount: BigNumber, token: TokenModel) => updateBalance(token, true, amount)

export const balanceMinus = (amount: BigNumber, token: TokenModel) => updateBalance(token, false, amount)

export const updateDeposit = (amount: BigNumber, isCredited: ?boolean) => ({
  type: WALLET_TIME_DEPOSIT,
  isCredited,
  amount,
})

export const depositPlus = (amount: BigNumber) => updateDeposit(amount, true)

export const depositMinus = (amount: BigNumber) => updateDeposit(amount, false)

export const allowance = (token: TokenModel, value: BigNumber, spender) => ({
  type: WALLET_ALLOWANCE,
  token,
  value,
  spender,
})

export const watchTransfer = (notice: TransferNoticeModel) => async (dispatch, getState) => {
  const tx: TxModel = notice.tx()
  const token: TokenModel = getState().get(DUCK_MAIN_WALLET).tokens().get(tx.symbol())

  dispatch(updateBalance(token, tx.isCredited(), tx.value()))

  const timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const timeHolderWalletAddress = await timeHolderDAO.getWalletAddress()
  let updateTIMEAllowance = false
  if (tx.to() === timeHolderWalletAddress) {
    dispatch(depositPlus(tx.value()))
    updateTIMEAllowance = true
  } else if (tx.from() === timeHolderWalletAddress) {
    dispatch(depositMinus(tx.value()))
    updateTIMEAllowance = true
  }
  if (updateTIMEAllowance) {
    const dao = await token.dao()
    dispatch(allowance(token, await dao.getAccountAllowance(timeHolderWalletAddress), timeHolderWalletAddress))
  }

  dispatch(notify(notice))
  dispatch({ type: WALLET_TRANSACTION, tx })
}

export const watchBalance = ({ symbol, balance /* balance3, balance6 */ }) => async (dispatch, getState) => {
  const token: TokenModel = getState().get(DUCK_MAIN_WALLET).tokens().get(symbol)
  dispatch(setBalance(token, balance))
}

export const watchInitWallet2 = () => (dispatch, getState) => {
  tokenService.on(EVENT_NEW_TOKEN, (token: TokenModel) => {
    const { profile } = getState().get(DUCK_SESSION)

    console.log('--actions4444#', token.toJS(), profile.toJS())
  })
}

export const watchInitWallet = () => async (dispatch, getState) => {
  const state = getState()

  const profile: ProfileModel = state.get(DUCK_SESSION).profile
  const previous = state.get(DUCK_MAIN_WALLET).tokens()

  dispatch({ type: WALLET_TOKENS_FETCH })
  const dao = await contractsManagerDAO.getERC20ManagerDAO()
  let tokens = await dao.getUserTokens(profile.tokens().toArray())
  dispatch({ type: WALLET_TOKENS, tokens })
  dispatch(getAccountTransactions(tokens))

  const toStopArray = previous.filter((k) => !tokens.get(k)).valueSeq().toArray().map((token: TokenModel) => {
    const dao = token.dao()
    return dao.stopWatching()
  })
  if (toStopArray.length) {
    await Promise.all(toStopArray)
  }

  const timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const [ timeHolderAddress, timeHolderWalletAddress ] = await Promise.all([
    timeHolderDAO.getAddress(),
    timeHolderDAO.getWalletAddress(),
  ])

  let contractNames = {}
  contractNames[ timeHolderAddress ] = TIME + ' Holder'
  ApprovalNoticeModel.setContractNames(contractNames)
  dispatch({ type: WALLET_TIME_ADDRESS, address: timeHolderWalletAddress })

  // NOTE @ipavlenko: BCC and BTC addresses usually the same.
  // Decided to manage them independently to simplify further works on multiple wallets. .
  dispatch({ type: WALLET_BTC_ADDRESS, address: btcProvider.getAddress() })
  dispatch({ type: WALLET_BCC_ADDRESS, address: bccProvider.getAddress() })
  dispatch({ type: WALLET_NEM_ADDRESS, address: nemProvider.getAddress() })

  tokens = tokens.filter((k) => !previous.get(k)).valueSeq().toArray()
  for (let token: TokenModel of tokens) {
    dispatch(addMarketToken(token.symbol()))
    const dao = token.dao()
    await dao.watchTransfer((notice) => dispatch(watchTransfer(notice)))
    if (dao.watchBalance) {
      await dao.watchBalance((balance) => dispatch(watchBalance(balance)))
    }
    await dao.watchApproval((notice: ApprovalNoticeModel) => {
      dispatch({ type: WALLET_ALLOWANCE, token, value: notice.value(), spender: notice.spender() })
      dispatch(notify(notice.setToken(token)))
    })
  }
}

export const mainTransfer = (token: TokenModel, amount: string, recipient) => async (dispatch) => {
  const amountLocal = new BigNumber(amount)

  dispatch(balanceMinus(amountLocal, token))
  // TODO @bshevchenko: sub balances with values of outcome pending transactions
  try {
    const dao = await token.dao()
    await dao.transfer(recipient, amountLocal)
  } finally {
    // compensation for update in watchTransfer
    dispatch(balancePlus(amountLocal, token))
  }
}

export const mainApprove = (token: TokenModel, amount: Number, spender: string) => async () => {
  try {
    const amountBN = new BigNumber(amount)
    const dao = await token.dao()
    await dao.approve(spender, amountBN)
  } catch (e) {
    // no rollback
    // eslint-disable-next-line
    console.error('approve error', e.message)
  }
}

export const depositTIME = (amount: string) => async (dispatch, getState) => {
  const amountBN = new BigNumber(amount)
  const token: TokenModel = getState().get(DUCK_MAIN_WALLET).tokens().get(TIME)

  dispatch(balanceMinus(amountBN, token))

  try {
    const dao = await contractsManagerDAO.getTIMEHolderDAO()
    await dao.deposit(amountBN)
  } finally {
    // compensation for update in watchTransfer
    dispatch(balancePlus(amountBN, token))
  }
}

export const withdrawTIME = (amount: string) => async (dispatch) => {
  const amountBN = new BigNumber(amount)

  dispatch(depositMinus(amountBN))

  try {
    const dao = await contractsManagerDAO.getTIMEHolderDAO()
    await dao.withdraw(amountBN)
  } finally {
    dispatch(depositPlus(amountBN))
  }
}

export const initTIMEDeposit = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getTIMEHolderDAO()
  const deposit = await dao.getAccountDepositBalance()
  dispatch(updateDeposit(deposit, null))
}

export const updateIsTIMERequired = () => async (dispatch) => {
  dispatch({ type: WALLET_IS_TIME_REQUIRED, value: await assetDonatorDAO.isTIMERequired() })
}

export const requireTIME = () => async (dispatch) => {
  try {
    await assetDonatorDAO.requireTIME()
  } catch (e) {
    // no rollback
    // eslint-disable-next-line
    console.error('require time error', e.message)
  }
  await dispatch(updateIsTIMERequired())
}

/**
 * LATEST TRANSACTIONS
 */
const getTransferId = 'wallet'
let lastCacheId
let txsCache = []

export const getAccountTransactions = (tokens) => async (dispatch) => {
  dispatch({ type: WALLET_TRANSACTIONS_FETCH })

  const tokensLocal = tokens.valueSeq().toArray()

  const cacheId = Object.values(tokensLocal).map((v: TokenModel) => v.symbol()).join(',')

  const reset = lastCacheId && cacheId !== lastCacheId
  lastCacheId = cacheId
  if (reset) {
    txsCache = []
  }

  let txs = txsCache.slice(0, TXS_PER_PAGE)
  txsCache = txsCache.slice(TXS_PER_PAGE)

  if (txs.length < TXS_PER_PAGE) { // so cache is empty
    const promises = []
    for (let token: TokenModel of tokensLocal) {
      if (reset) {
        token.dao().resetFilterCache()
      }
      promises.push(token.dao().getTransfer(getTransferId))
    }
    const result = await Promise.all(promises)

    let newTxs = []
    for (let pack of result) {
      newTxs = [ ...newTxs, ...pack ]
    }

    newTxs.sort((a, b) => b.get('time') - a.get('time'))

    txs = [ ...txs, ...newTxs ]
    txsCache = txs.slice(TXS_PER_PAGE)
    txs = txs.slice(0, TXS_PER_PAGE)
  }

  let map = new Immutable.Map()
  for (let tx: TxModel of txs) {
    map = map.set(tx.id(), tx)
  }

  dispatch({ type: WALLET_TRANSACTIONS, map })
}
