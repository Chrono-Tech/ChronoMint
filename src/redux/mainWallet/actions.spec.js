import BigNumber from 'bignumber.js'
import Immutable from 'immutable'
import { mockStore, store, accounts } from 'specsInit'

import contractsManagerDAO from 'dao/ContractsManagerDAO'
import ethereumDAO from 'dao/EthereumDAO'
import lhtDAO from 'dao/LHTDAO'
import type TIMEHolderDAO from 'dao/TIMEHolderDAO'

import ProfileModel from 'models/ProfileModel'
import TokenModel from 'models/TokenModel'
import TransferNoticeModel from 'models/notices/TransferNoticeModel'
import TxModel from 'models/TxModel'

import web3Provider from 'network/Web3Provider'

import { notify } from 'redux/notifier/actions'

import web3Converter from 'utils/Web3Converter'

import * as a from './actions'

const account1 = accounts[6]
const account2 = accounts[7]
let eth: TokenModel
let time: TokenModel
let lht: TokenModel

const amountToDeposit = '3.9438273'
const amountToDepositBN = new BigNumber(amountToDeposit)

const amountToTransferETH = '1.57483625'
const amountToTransferETHBN = new BigNumber(amountToTransferETH)

let requiredTIMEBN

const amountToTransferTIME = '2.0438572'
const amountToTransferTIMEBN = new BigNumber(amountToTransferTIME)

const amountToWithdraw = '0.378426'
const amountToWithdrawBN = new BigNumber(amountToWithdraw)

let timeHolderDAO: TIMEHolderDAO
let timeHolderAddress

let getStateTIME

describe('wallet actions', () => {
  beforeAll(() => {
    contractsManagerDAO.setAccount(account1)
  })

  it('should init watch', async () => {
    const get = (profileTokens = [], tokens = new Immutable.Map()) => (key) => {
      if (key === 'session') {
        return { profile: new ProfileModel({ tokens: profileTokens }) }
      }
      if (key === 'wallet') {
        return { tokens }
      }
    }
    let myStore = mockStore({ get: get([await lhtDAO.getAddress()]) })

    await myStore.dispatch(a.watchInitWallet())

    const tokens = myStore.getActions()[1].tokens
    eth = tokens.get(a.ETH)
    time = tokens.get(a.TIME)
    lht = tokens.get(lhtDAO.getSymbol())

    expect(eth instanceof TokenModel).toBeTruthy()
    expect(time instanceof TokenModel).toBeTruthy()
    expect(lht instanceof TokenModel).toBeTruthy()

    expect(eth.balance()).toEqual(new BigNumber(100))

    // test stop watching of previous tokens
    expect(lht.dao().getWatchedEvents().length).toEqual(3) // 1 approval + 2 transfer (income and outcome)
    myStore = mockStore({ get: get([], tokens) })
    await myStore.dispatch(a.watchInitWallet())
    expect(lht.dao().getWatchedEvents().length).toEqual(0)
  })

  it('should transfer ETH', async () => {
    const account1Balance = await ethereumDAO.getAccountBalance(account1)
    const account2Balance = await ethereumDAO.getAccountBalance(account2)

    expect(account1Balance.toNumber()).toEqual(100)
    expect(account2Balance.toNumber()).toEqual(100)

    await store.dispatch(a.transfer(eth, amountToTransferETH, account2))

    expect(store.getActions()).toEqual([
      a.balanceMinus(amountToTransferETHBN, eth),
      a.balancePlus(amountToTransferETHBN, eth),
    ])

    const latestBlock = await web3Provider.getBlock('latest')
    const tx = await web3Provider.getTransaction(latestBlock.transactions[0])
    const gasUsed = new BigNumber(latestBlock.gasUsed)
    const gasFee = new BigNumber(web3Converter.fromWei(gasUsed.mul(tx.gasPrice)))

    expect(await ethereumDAO.getAccountBalance()).toEqual(account1Balance.minus(amountToTransferETHBN).minus(gasFee))
    expect(await ethereumDAO.getAccountBalance(account2)).toEqual(account2Balance.plus(amountToTransferETHBN))
  })

  it('should require TIME', async () => {
    const balanceBefore = await time.dao().getAccountBalance()
    expect(balanceBefore.toNumber()).toEqual(0)

    await store.dispatch(a.requireTIME())

    expect(store.getActions()).toEqual([
      { type: a.WALLET_IS_TIME_REQUIRED, value: true },
    ])

    const balanceAfter = await time.dao().getAccountBalance()
    expect(balanceAfter.toNumber()).toEqual(10)
  })

  it('should transfer ERC20 tokens', async () => {
    const account1Balance = await time.dao().getAccountBalance()
    const account2Balance = await time.dao().getAccountBalance(account2)

    expect(account1Balance.toNumber()).toEqual(10)
    expect(account2Balance.toNumber()).toEqual(0)

    time = time.setBalance(account1Balance)
    requiredTIMEBN = account1Balance

    await store.dispatch(a.transfer(time, amountToTransferTIMEBN, account2))

    expect(store.getActions()).toEqual([
      a.balanceMinus(amountToTransferTIMEBN, time),
      a.balancePlus(amountToTransferTIMEBN, time),
    ])

    expect(await time.dao().getAccountBalance()).toEqual(account1Balance.minus(amountToTransferTIMEBN))
    expect(await time.dao().getAccountBalance(account2)).toEqual(account2Balance.plus(amountToTransferTIMEBN))
  })

  it('should approve ERC20 tokens', async () => {
    timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
    timeHolderAddress = await timeHolderDAO.getWalletAddress()
    await store.dispatch(a.approve(time, amountToDeposit, timeHolderAddress))

    expect(await time.dao().getAccountAllowance(timeHolderAddress)).toEqual(new BigNumber(amountToDeposit))
  })

  it('should deposit TIME and init TIME deposit', async () => {
    getStateTIME = {
      get: (key) => {
        if (key === 'wallet') {
          return { tokens: new Immutable.Map({ [a.TIME]: time }) }
        }
      },
    }

    const myStore = mockStore(getStateTIME)

    await myStore.dispatch(a.depositTIME(amountToDeposit))
    await myStore.dispatch(a.initTIMEDeposit(amountToDeposit))

    expect(myStore.getActions()).toEqual([
      a.balanceMinus(amountToDepositBN, time),
      a.balancePlus(amountToDepositBN, time),
      a.updateDeposit(amountToDepositBN, null),
    ])
  })

  it('should withdraw TIME', async () => {
    await store.dispatch(a.withdrawTIME(amountToWithdraw))

    expect(store.getActions()).toEqual([
      a.depositMinus(amountToWithdrawBN),
      a.depositPlus(amountToWithdrawBN),
    ])

    expect(await timeHolderDAO.getAccountDepositBalance()).toEqual(amountToDepositBN.minus(amountToWithdrawBN))
  })

  it('should watch transfer, update balance & deposit & allowance, notify and add tx to list', async () => {
    const isCredited = false
    const value = new BigNumber('1.483729')
    const tx = new TxModel({
      to: timeHolderAddress,
      credited: isCredited,
      value,
      symbol: time.symbol(),
    })
    const notice = new TransferNoticeModel({ tx })

    const myStore = mockStore(getStateTIME)
    await myStore.dispatch(a.watchTransfer(notice))

    expect(myStore.getActions()).toEqual([
      a.updateBalance(time, isCredited, value),
      a.depositPlus(value),
      a.allowance(time, new BigNumber(0), timeHolderAddress),
      notify(notice),
      { type: a.WALLET_TRANSACTION, tx },
    ])
  })

  it('should get transactions by account', async () => {
    await store.dispatch(a.getAccountTransactions(new Immutable.Map({
      [a.ETH]: eth,
      [a.TIME]: time,
      [lhtDAO.getSymbol()]: lht,
    })))

    const txs: Array<TxModel> = store.getActions()[1].map.valueSeq().toArray()

    expect(txs[0].value()).toEqual(amountToWithdrawBN)
    expect(txs[1].value()).toEqual(amountToDepositBN)
    expect(txs[2].value()).toEqual(amountToTransferTIMEBN)
    expect(txs[3].value()).toEqual(requiredTIMEBN)
    expect(txs[4].value()).toEqual(amountToTransferETHBN)
  })

  afterAll(() => {
    contractsManagerDAO.setAccount(accounts[0])
  })
})
