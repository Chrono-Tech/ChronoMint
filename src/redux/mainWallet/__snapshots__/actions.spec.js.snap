// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wallet actions should approve ERC20 tokens 1`] = `"3.9438273"`;

exports[`wallet actions should check allowance 1`] = `
Object {
  "spender": "",
  "token": Immutable.Record {
    "id": null,
    "timestamp": null,
    "isFetching": false,
    "isFetched": false,
    "transactionHash": null,
    "isPending": false,
    "isFailed": false,
    "dao": null,
    "address": null,
    "decimals": null,
    "name": null,
    "symbol": null,
    "balance": "0",
    "allowance": Immutable.Map {},
    "url": null,
    "icon": null,
    "fee": null,
    "withFee": null,
    "platform": null,
    "totalSupply": "0",
    "managersList": null,
    "isReissuable": null,
    "isOptional": true,
    "feeAddress": null,
  },
  "type": "mainWallet/ALLOWANCE",
  "value": "100",
}
`;

exports[`wallet actions should check balanceMinus 1`] = `
Object {
  "amount": "100",
  "isCredited": true,
  "token": Immutable.Record {
    "id": null,
    "timestamp": null,
    "isFetching": false,
    "isFetched": false,
    "transactionHash": null,
    "isPending": false,
    "isFailed": false,
    "dao": null,
    "address": null,
    "decimals": null,
    "name": null,
    "symbol": null,
    "balance": "0",
    "allowance": Immutable.Map {},
    "url": null,
    "icon": null,
    "fee": null,
    "withFee": null,
    "platform": null,
    "totalSupply": "0",
    "managersList": null,
    "isReissuable": null,
    "isOptional": true,
    "feeAddress": null,
  },
  "type": "mainWallet/BALANCE",
}
`;

exports[`wallet actions should check balancePlus 1`] = `
Object {
  "amount": "100",
  "isCredited": true,
  "token": Immutable.Record {
    "id": null,
    "timestamp": null,
    "isFetching": false,
    "isFetched": false,
    "transactionHash": null,
    "isPending": false,
    "isFailed": false,
    "dao": null,
    "address": null,
    "decimals": null,
    "name": null,
    "symbol": null,
    "balance": "0",
    "allowance": Immutable.Map {},
    "url": null,
    "icon": null,
    "fee": null,
    "withFee": null,
    "platform": null,
    "totalSupply": "0",
    "managersList": null,
    "isReissuable": null,
    "isOptional": true,
    "feeAddress": null,
  },
  "type": "mainWallet/BALANCE",
}
`;

exports[`wallet actions should check depositMinus 1`] = `
Object {
  "amount": "100",
  "isCredited": false,
  "type": "mainWallet/TIME_DEPOSIT",
}
`;

exports[`wallet actions should check depositPlus 1`] = `
Object {
  "amount": "100",
  "isCredited": true,
  "type": "mainWallet/TIME_DEPOSIT",
}
`;

exports[`wallet actions should check setBalance 1`] = `
Object {
  "amount": "100",
  "token": Immutable.Record {
    "id": null,
    "timestamp": null,
    "isFetching": false,
    "isFetched": false,
    "transactionHash": null,
    "isPending": false,
    "isFailed": false,
    "dao": null,
    "address": null,
    "decimals": null,
    "name": null,
    "symbol": null,
    "balance": "0",
    "allowance": Immutable.Map {},
    "url": null,
    "icon": null,
    "fee": null,
    "withFee": null,
    "platform": null,
    "totalSupply": "0",
    "managersList": null,
    "isReissuable": null,
    "isOptional": true,
    "feeAddress": null,
  },
  "type": "mainWallet/BALANCE_SET",
}
`;

exports[`wallet actions should check updateBalance 1`] = `
Object {
  "amount": "100",
  "isCredited": false,
  "token": Immutable.Record {
    "id": null,
    "timestamp": null,
    "isFetching": false,
    "isFetched": false,
    "transactionHash": null,
    "isPending": false,
    "isFailed": false,
    "dao": null,
    "address": null,
    "decimals": null,
    "name": null,
    "symbol": null,
    "balance": "0",
    "allowance": Immutable.Map {},
    "url": null,
    "icon": null,
    "fee": null,
    "withFee": null,
    "platform": null,
    "totalSupply": "0",
    "managersList": null,
    "isReissuable": null,
    "isOptional": true,
    "feeAddress": null,
  },
  "type": "mainWallet/BALANCE",
}
`;

exports[`wallet actions should check updateDeposit 1`] = `
Object {
  "amount": false,
  "isCredited": "100",
  "type": "mainWallet/TIME_DEPOSIT",
}
`;

exports[`wallet actions should deposit TIME and init TIME deposit 1`] = `
Array [
  Object {
    "amount": "0",
    "isCredited": null,
    "type": "mainWallet/TIME_DEPOSIT",
  },
]
`;

exports[`wallet actions should get transactions by account 1`] = `"10"`;

exports[`wallet actions should init watch 1`] = `true`;

exports[`wallet actions should init watch 2`] = `true`;

exports[`wallet actions should init watch 3`] = `true`;

exports[`wallet actions should init watch 4`] = `"100"`;

exports[`wallet actions should init watch 5`] = `3`;

exports[`wallet actions should init watch 6`] = `0`;

exports[`wallet actions should require TIME 1`] = `0`;

exports[`wallet actions should require TIME 2`] = `
Array [
  Object {
    "type": "mainWallet/IS_TIME_REQUIRED",
    "value": true,
  },
]
`;

exports[`wallet actions should require TIME 3`] = `10`;

exports[`wallet actions should transfer ERC20 tokens 1`] = `10`;

exports[`wallet actions should transfer ERC20 tokens 2`] = `0`;

exports[`wallet actions should transfer ETH 1`] = `100`;

exports[`wallet actions should transfer ETH 2`] = `100`;
